### PR TITLE
Consistently raise an `ArgumentError` if the `ActiveSupport::Cache` key is blank

### PR DESCRIPTION
**Pull request by joshuay03**
### Motivation / Background

Currently, when the expanded cache key is `nil` or `""`:
- An `ArgumentError` is raised in the `mem_cache` store implementation which happens within the client here: https://github.com/petergoldstein/dalli/blob/a2e2951ec8de2ba6fb00bd4e0775f9846f1e99b9/lib/dalli/key_manager.rb#L50-L55.

- A `TypeError` is raised **only for** `nil` in the `redis` store implementation which happens within the client here: https://github.com/redis-rb/redis-client/blob/54f82471ce8b9fb06f03d7b1d4a0710cd58e6524/lib/redis_client/command_builder.rb#L75.

- The `file_store` implementation would error as such:
   Screenshot 2023-04-24 at 3 32 11 pm

- The `memory` and `null` stores would succeed with the operation without raising.

This PR ensures that we consistently raise on a blank key for all store implementations.

### Detail

This Pull Request adds a validation step in `ActiveSupport::Cache::Store#normalize_key` to ensure that an `ArgumentError` is consistently raised when the expanded cache key is `nil` or `""` within all methods relying on `#read_entry`, `#write_entry` and `#delete_entry` (`#read`, `#fetch` etc.).

### Additional information

I'm unsure if we should raise on a blank or just a `nil` key. I went with the former seeing as it already happens within one implementation.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activesupport/lib/active_support/cache.rb [hunk 1]
```diff
@@ -817,6 +817,12 @@ def namespace_key(key, options = nil)
           end
         end
 
+        # Validates the expanded key.
+        # Raises an exception when the key is +nil+ or an empty string.
+        def validated_key(key)
+          key.presence || raise(ArgumentError, "key cannot be blank")
+        end
```
Comment (byroot) [inline activesupport/lib/active_support/cache.rb:824]: I don't think this really deserve an extracted method. I'd suggest poutting this logic in `normalize_key`.

File: activesupport/test/cache/cache_key_test.rb [hunk 2]
```diff
@@ -78,6 +78,13 @@ def test_expand_cache_key_of_array_like_object
     assert_equal "foo/bar/baz", ActiveSupport::Cache.expand_cache_key(%w{foo bar baz}.to_enum)
   end
 
+  def test_validate_key
+    store = ActiveSupport::Cache::Store.new
+    assert_raises(ArgumentError) { store.send(:validate_key, nil) }
+    assert_raises(ArgumentError) { store.send(:validate_key, "") }
+    assert_nothing_raised { store.send(:validate_key, "foo_key") }
```
Comment (byroot) [inline activesupport/test/cache/cache_key_test.rb:85]: I don't think we need to call private APIs to test that behavior.

**Review by byroot — CHANGES_REQUESTED** [review]
Some small comments, but generally in favor of this.

**Comment by ghiculescu** [comment]
@joshuay03 could you add the tests to https://github.com/rails/rails/blob/main/activesupport/test/cache/behaviors/cache_store_behavior.rb rather than duplicating them in every test class?

File: activesupport/lib/active_support/cache.rb (hunk 1 above)
Comment (joshuay03) [reply activesupport/lib/active_support/cache.rb]: Thanks @byroot I've moved it into `normalize_key`.

File: activesupport/test/cache/cache_key_test.rb (hunk 2 above)
Comment (joshuay03) [reply activesupport/test/cache/cache_key_test.rb]: 👍🏽 I've removed this.

**Comment by joshuay03** [comment]
> @joshuay03 could you add the tests to [`main`/activesupport/test/cache/behaviors/cache_store_behavior.rb](https://github.com/rails/rails/blob/main/activesupport/test/cache/behaviors/cache_store_behavior.rb?rgh-link-date=2023-04-26T04%3A17%3A00Z) rather than duplicating them in every test class?

Updated to be a single cache store behaviour test!

**Comment by joshuay03** [comment]
Currently failing some tests that explicitly set the key to be blank. I think we just need to remove these...

File: activesupport/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+*   Consistently raise an `ArgumentError` if the cache key is blank.
```
Comment (byroot) [inline activesupport/CHANGELOG.md:1]: Maybe mention that it has to do with ActiveSupport::Cache.

File: activesupport/lib/active_support/cache.rb [hunk 4]
```diff
@@ -783,10 +783,14 @@ def normalize_options(options)
...
 
-        # Expands and namespaces the cache key. May be overridden by
-        # cache stores to do additional normalization.
+        # Expands and namespaces the cache key.
+        # Raises an exception when the key is +nil+ or an empty string.
+        # May be overridden by cache stores to do additional normalization.
         def normalize_key(key, options = nil)
-          namespace_key expanded_key(key), options
+          str_key = expanded_key(key)
+          raise(ArgumentError, "key cannot be blank") unless str_key&.length&.positive?
```
[the remark this answers is not in the document]
Comment (joshuay03) [reply activesupport/lib/active_support/cache.rb]: Actioned!

File: activesupport/CHANGELOG.md (hunk 3 above)
Comment (joshuay03) [reply activesupport/CHANGELOG.md]: Updated!

**Review by byroot — APPROVED** [review]

**Comment by joshuay03** [comment]
Cheers @byroot

**Comment by georgeclaghorn** [comment]
Corner case: this appears to have broken [collection caching](https://edgeguides.rubyonrails.org/caching_with_rails.html#collection-caching) when the collection is empty and the log level is `debug`.

`keyed_collection` is an empty hash here, so `keyed_collection.keys` is `[]` and `ActiveSupport::Cache::Store#read_multi` is called with no args:

https://github.com/rails/rails/blob/490804f7d5dd724c91c6b19f8c0fcc71a07df616/actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb#L34

`read_multi` passes its args to `instrument` as though they were one key:

https://github.com/rails/rails/blob/490804f7d5dd724c91c6b19f8c0fcc71a07df616/activesupport/lib/active_support/cache.rb#L439-L443

It’s the `normalize_key` call in `instrument` which now blows up:

https://github.com/rails/rails/blob/490804f7d5dd724c91c6b19f8c0fcc71a07df616/activesupport/lib/active_support/cache.rb#L866-L869

**Comment by joshuay03** [comment]
> Corner case: this appears to have broken [collection caching](https://edgeguides.rubyonrails.org/caching_with_rails.html#collection-caching) when the collection is empty and the log level is `debug`.

Thanks! @georgeclaghorn I'll get a fix up for this.

**Comment by georgeclaghorn** [comment]
Here’s a failing test:

```diff
diff --git a/actionview/test/template/render_test.rb b/actionview/test/template/render_test.rb
index 524065921f..46becb3e2a 100644
--- a/actionview/test/template/render_test.rb
+++ b/actionview/test/template/render_test.rb
@@ -869,6 +869,7 @@ class CachedCustomer < Customer; end
     assert_equal ActionView::FileSystemResolver, view_paths.first.class
 
     ActionView::PartialRenderer.collection_cache = ActiveSupport::Cache::MemoryStore.new
+    ActionView::PartialRenderer.collection_cache.logger = Logger.new(nil, level: :debug)
 
     setup_view(view_paths)
   end
@@ -968,6 +969,10 @@ def b.to_partial_path; "test/partial_iteration_2"; end
     ], splited_result
   end
 
+  test "collection caching with empty collection" do
+    assert_nil @view.render(partial: "test/cached_customer", collection: [], cached: true)
+  end
+
   private
     def cache_key(*names, virtual_path)
       digest = ActionView::Digestor.digest name: virtual_path, format: :html, finder: @view.lookup_context, dependencies: []
```

```
Error:
CachedCollectionViewRenderTest#test_collection_caching_with_empty_collection:
ArgumentError: key cannot be blank
    /Users/george/Work/rails/rails/activesupport/lib/active_support/cache.rb:801:in `normalize_key'
    /Users/george/Work/rails/rails/activesupport/lib/active_support/cache.rb:868:in `instrument'
    /Users/george/Work/rails/rails/activesupport/lib/active_support/cache.rb:443:in `read_multi'
    /Users/george/Work/rails/rails/actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb:34:in `cache_collection_render'
    /Users/george/Work/rails/rails/actionview/lib/action_view/renderer/collection_renderer.rb:169:in `block in render_collection'
    /Users/george/Work/rails/rails/activesupport/lib/active_support/notifications.rb:208:in `instrument'
    /Users/george/Work/rails/rails/actionview/lib/action_view/renderer/collection_renderer.rb:155:in `render_collection'
    /Users/george/Work/rails/rails/actionview/lib/action_view/renderer/collection_renderer.rb:127:in `render_collection_with_partial'
    /Users/george/Work/rails/rails/actionview/lib/action_view/renderer/renderer.rb:74:in `render_partial_to_object'
    /Users/george/Work/rails/rails/actionview/lib/action_view/renderer/renderer.rb:29:in `render_to_object'
    /Users/george/Work/rails/rails/actionview/lib/action_view/renderer/renderer.rb:24:in `render'
    /Users/george/Work/rails/rails/actionview/lib/action_view/helpers/rendering_helper.rb:37:in `block in render'
    /Users/george/Work/rails/rails/actionview/lib/action_view/base.rb:291:in `in_rendering_context'
    /Users/george/Work/rails/rails/actionview/lib/action_view/helpers/rendering_helper.rb:33:in `render'
    /Users/george/Work/rails/rails/actionview/test/template/render_test.rb:973:in `block in <class:CachedCollectionViewRenderTest>'


bin/test test/template/render_test.rb:972
```

**Comment by joshuay03** [comment]
I've got a PR up to fix this: https://github.com/rails/rails/pull/48145.

**Pull request merged**